cpp/include/cudf/aggregation.hpp

```diff
@@ -599,6 +609,255 @@ std::unique_ptr<Base> make_udf_aggregation(udf_type type,
                                            std::string const& user_defined_aggregator,
                                            data_type output_type);
 
+/**
+ * @brief The interface for host-based UDF implementation.
+ *
+ * An implementation of host-based UDF needs to be derived from this base class, defining
+ * its own version of the required functions. In particular:
+ *  - The derived class is required to implement `get_empty_output`, `operator()`, `is_equal`,
+ *    and `clone` functions.
+ *  - If necessary, the derived class can also override `do_hash` to compute hashing for its
+ *    instance, and `get_required_data` to selectively access to the input data as well as
+ *    intermediate data provided by libcudf.
+ */
+struct host_udf_base {
```

Comment:
Would it be possible to isolate some of this into a separate header? This header is included in many places and this feature is very specific. Changes to it would require re-compiling alot of source files unnecessarily.

Reply:
Done. Thanks again for the idea. Now whenever modifying `host_udf_base`, the number of recompiled files reduces from 161 to 7, including test and example.